### ISSUE TITLE
Spec-decode draft undertrained: 0% acceptance (20 training samples) — needs a real dataset

### ISSUE THREAD

**Issue by bong-water-water-bong** [1bit-MONSTER/1bit-MONSTER#1509]
## Status (from #1487 close, 2026-08-05)

The pipeline is proven end-to-end: ROCm 7.2 torch venv rebuilt (`~/spec-decode/train-venv`), target model downloaded, on-policy data regenerated, target cache prepared, Eagle3 draft trained (10 epochs), **`spec-decode/checkpoints/eagle3_draft_v2.bin` exported (1.3 GB)**. `npu_engine_spec models/qwen3_0_6b.q4nx spec-decode/checkpoints/eagle3_draft_v2.bin 4` runs: draft loads, proposes, verifies.

## The gap

**0/12 proposals accepted (0%)** — the draft is undertrained: only **20 valid training samples** existed locally (gsm8k + mtbench regen, `~/DeepSpec/train_datasets/qwen3_0.6b/`), loss 22.1. The batching unlock (#1485's batched-decode path) needs an acceptance rate worth verifying against.

## What's needed

1. **More/better training data** — the regen path (`spec-decode/scripts_local/regen_train_data.py` → `prepare_target_cache.py`) works; run it on a larger corpus, or wire DeepSpec's data pipeline for 0.6B (much smaller than the 4B 38 TB case)
2. Retrain (`/home/bcloud/tmp/train_job.sh` or `spec-decode/train_draft.py`), re-export
3. Target: acceptance ≥ 50% → then the spec-decode speedup claims (1.5-2.5×) become measurable

## Repro

```
export TMPDIR=/home/bcloud/tmp
export PYTHONPATH=/home/bcloud/DeepSpec:/home/bcloud/projects/1bit-systems/spec-decode
/home/bcloud/spec-decode/train-venv/bin/torchrun --nproc_per_node=1 train.py --config spec-decode/configs/eagle3_qwen3_0.6b.py
```

**Comment by bong-water-water-bong** [comment]
## Real dataset + working trainer — draft loss 22.1 → 0.32, acceptance proxy 0% → 42.5% (merged in PR #1598, commit 86824502)

Premise check: `~/DeepSpec`, `~/spec-decode/train-venv` and the old checkpoints did **not** exist on this machine (ROCm-torch venv absent; system torch is CPU-only). So instead of patching the old pipeline, a self-contained trainer was written that reproduces the deployed draft 1:1 from `engine/npu/src/npu_engine_spec.hip` (the C++ is the spec):

- **Dataset:** `spec-decode/train_data/gsm8k_train.jsonl` — **7,473 samples** (openai/gsm8k train split), turns format, parse-validated. 3.9 MB.
- **Trainer:** `spec-decode/train_eagle3.py` (CPU, no new deps) — frozen Qwen3-0.6B trunk, 1-layer draft conditioned on hidden states at target layers [1,6,12,18,24], 7-token teacher-forced windows, tied/frozen LM head. Writes the exact 16-tensor f32 layout the C++ reads.
- **Result:** loss **22.1 → 0.32** (mean last-25, plateau ~sample 150). Acceptance proxy (draft argmax == target argmax, the C++'s exact criterion): **42.5% mean** over 9 eval prompts (26–65%; 1.2% outlier is a 1382-token OOD Malayalam translation). Was 0/12.
- **Checkpoint:** `spec-decode/checkpoints/eagle3_draft_v3.bin` (1.35 GB, exact layout size).
- **Runnable check:** `spec-decode/tests/test_train_eagle3.py` (dataset format + 2-sample train + .bin layout vs C++ reader) — all pass.

**Caveats for the hardware owner:**
1. `npu_engine_spec.hip` `draft_forward` has an **OOB read**: `nr(d_attn, qN, NH*HD)` indexes 2047/1023 but only 128 floats are uploaded → garbage norm weights on the real NPU. The trainer uses the intended per-head broadcast; if real-NPU acceptance is still low, fix that kernel path first (needs a per-head norm kernel — unverifiable here, left unfixed).
2. Real-NPU acceptance couldn't be measured on this box (no XRT/xclbin/ROCm-torch). 42.5% is the exact-argmax proxy.
3. Full 7,473-sample epoch ≈ 4.5 h CPU projected; the short run (300 samples, 1 epoch) proves the pipeline and trend. Run on a ROCm box to train to convergence.

**Remaining steps:** fix the draft_forward OOB read, then verify acceptance on the real NPU.

**Comment by bong-water-water-bong** [comment]
## draft_forward OOB read fixed + GPU-validated on real hardware (commit 08121d2f)

The caveat from the trainer PR is now fixed on the NPU box itself:

- **Fix:** per-head qk-norm loop in both engines that had the bug — `npu_engine_spec.hip` (draft Q at NH=16 heads, K at NKV=8) and `npu_engine_gpurender.hip` (same pattern, prefill + decode paths). Weights are HD=128 floats; the old code normed NH*HD=2048 elements against them (OOB read + wrong grouping). Matches the trainer's per-head broadcast exactly.
- **Validation:** `engine/npu/tests/test_qknorm_perhead.hip` on gfx1151 (TheRock ROCm): GPU per-head pattern vs CPU per-head reference — **max abs err 0.0019** (f16 rounding), PASS.
- **Toolchain note:** TheRock gfx1151 nightly 7.14.0a20260612 installed at /opt/rocm-therock (old 10.1 nightly moved to /opt/rocm-therock-10.1nightly). **7.14's HSA runtime segfaults in GpuAgent::InitDma on kernel 7.0.0-29** — binaries must run with the 10.1 runtime via LD_LIBRARY_PATH (compiling with 7.14 is fine).

**Still blocked:** real-NPU acceptance measurement. npu_engine_spec's batched verify path floods AMD-Vi IO_PAGE_FAULTs and runs ~17s/layer (28-layer verify ≈ 8 min/token) — pre-existing bug, present with the NPU exclusively free, unrelated to this fix. Filed separately.

**Comment by bong-water-water-bong** [comment]
The 0%-acceptance harness artifact is resolved by ab076830 (spec-decode: fix the draft trunk mismatch (train/deploy)).

Two linked bugs pinned acceptance at 0% regardless of draft quality:
1. Harness fed the draft a degenerate trunk (one hidden state replicated 5x) instead of the real 5-layer states verify_batch fills — now uses trunk[].
2. Trainer window() used per-position trunk[i] (teacher-forcing-only); decode must evolve from block-start trunk — now uses trunk[w0], deployment-equivalent.

Also fixed the malformed prefill prompt. Retraining then proceeded through v7→v14 (step counts to s11500). Closing as resolved; residual draft-quality is a training-data project, tracked separately if revived.
